### ISSUE TITLE
Proposal of a drought duration climate index

### ISSUE THREAD

**Issue by yrouranos** [Ouranosinc/xclim#843]
* xclim version: 0.25.0
* Python version: 3.7.6
* Operating System: LinuxMint 20.1

### Description
A drought duration climate index was developed (currently available in /scen_workflow_afr/indices.py). The algorithm is currently being tested (using climate data from West Africa). This index can be used to identify the impacts of climate change in countries where a monsoon is present.

### What I Did
A call to the developed function returns an instance of `xarray.DataArray` which contains the number of days per year that are comprised in a dry period. 2 methods were implemented to define what a dry period is (based on a daily or total precipitation amount). Parameters allow adjusting detection rules in different countries or regions.

drydurtot ('1d' method): A dry period is a sequence of _D_threshold_ consecutive days during which the daily precipitation amount (_P_day_) is lower than a threshold value (_P_threshold_).

drydurtot ('tot' method): A dry period is a sequence of _D_threshold_ consecutive days during which the total amount of precipitation (_P_total_) is lower than a threshold value (_P_threshold_).

In both cases, the user can specify the days of year to consider (between day DOY_min and DOY_max). 

### What I Received
This section is not required.

**Comment by aulemahal** [comment]
How does this compare to : `dry_spell_total_length` ? It looks like this is similar to your `tot` method. Maybe, the `1d` method could be added to the existing function?
https://xclim.readthedocs.io/en/stable/indices.html?highlight=dry_spell#xclim.indices.dry_spell_total_length

It was added quite recently in xclim.

Currently, we don't have many methods with the equivalent of `DOY_min` and `DOY_max`. I would try to think of a more general way to add this functionality. It is kinda similar in idea to the problem in #649 where we decided to not implement anything in xclim. Instead, users would mask the data themselves and disable/config missing checks accordingly.

**Comment by yrouranos** [comment]
@aulemahal It's indeed very similar to `dry_spell_total_length`. Adding the `1d` method to the existing function would probably be done easily by replacing `.sum() < thresh` with `.max() < thresh`.

Regarding `DOY_min` and `DOY_max`, I'm not sure how that would work with the mask for the simple reason that I'm not familiar with this concept in `xarray`.  I need the analysis to be performed on all days of the year, while compiling only the days between `DOY_min` and `DOY_max`. For instance, if we have at least `D_threshold` consecutive days with `P_day < P_threshold` (occurring from doy 100 to 120), but `(DOY_min==105) and (DOY_max==115)`, then the algorithm needs to compile 11 days (i.e., 115 - 105 + 1) out of the potential 21 days (i.e., 120 - 100 + 1). Do you think it could work with a mask?

**Comment by aulemahal** [comment]
The mask idea would simply be to put NaN on days 0 to 104, then the algorithm will be unable to find any dry spells before 105. And given that window <= 11, it would return 11 days as expected. Does that make sense?

**Comment by yrouranos** [comment]
@aulemahal My example was not complete. I forgot to assign a value to the parameter `D_threshold`. The two climate indices (`drydurtot` vs. masked `dry_spell_total_length`) would return different values if `D_threshold == 14`. With the current index (`drydurtot`), the number of compiled days would be 11 (because `21 > D_threshold`), but zero with `dry_spell_total_length` (because `11 < D_threshold`). So, I understand that masking would not work in that case.

I would not recommend adding the climate index `drydurtot` (or improved name) if there is already a very similar index in xclim (`dry_spell_total_length`), perhaps only the `1d` method. Otherwise, that would be very redundant, and maybe not compatible width dask. In a case like this, a possibility would be to keep the index in the workflow (not in xclim). However, if this type of functionality is often required, I would be in favour of finding a global xclim solution (related to all climate indices).

**Comment by aulemahal** [comment]
I see! Indeed my proposed masking is not appropriate.

We could implement it. 2 things
1. we have a few indices where year-agnostic dates are given as strings, ex: `01-04` (Jan 4th),  I'm guessing this would be ok to use, instead of doys? 
2. Because of the potential to reuse the date_min / date_max functionality, I suggest implementing some sort of `mask_between_dates` function (perharps in `core.calendar`?). For our specific case, if bounds are given, this would be called between the thresholding and the aggregation. Thus, spells are computed with the whole year, but only the days between the bounds are counted.

**Comment by yrouranos** [comment]
@aulemahal I attached the 30 unit tests that I used to verify the behaviour of an improved `dry_spell_total_length` climate index. I compared results between the current xclim algorithm (O) and the proposed algorithm (N), whenever a comparison was possible between the 2 algorithms (if the start_date and end_date did not impact results). This required adding the operator "max" in the existing xclim algorithm (locally). Each case was based on a 2-year artificial dataset. See the legend at the bottom of the spreadsheet. I don't always document my unit tests to this extent, but I thought it was the best way to document my observations. Let me know if you would like to talk about it.
[dry_spell_total_length.xlsx](https://github.com/Ouranosinc/xclim/files/7385123/dry_spell_total_length.xlsx)
